src/lib.rs

```diff
@@ -246,7 +259,156 @@ impl BinaryQF {
+        // 4. a' <- a/g
+        let a1 = &self.a / &g;
+        // 5-8. if a = b then t' <- 0 else t' <- t/g
+        let t1 = if self.a == self.b {
+            BigInt::zero()
+        } else {
+            &t / &g
+        };
+        // 9. b0 <- b mod g
+        let b0 = self.b.modulus(&g);
+        // 10. ε <- [b >= 0]
+        let e = self.b >= BigInt::zero();
+
+        let delta = self.discriminant();
+        Some(BinaryQFCompressed {
+            a1,
+            t1,
+            g,
+            b0,
+            e,
+            delta,
+        })
+    }
+
+    pub fn from_compressed(compressed: BinaryQFCompressed) -> Option<Self> {
+        Some(Self::binary_quadratic_form_disc(
+            &ABDeltaTriple::from_compressed(compressed)?,
+        ))
+    }
+}
+
+impl ABDeltaTriple {
+    pub fn from_compressed(compressed: BinaryQFCompressed) -> Option<Self> {
+        use arithmetic::A;
+
+        let BinaryQFCompressed {
+            a1,
+            t1,
+            g,
+            b0,
+            e,
+            delta,
+        } = compressed;
+
+        // 1. a <- g * a'
+        let a = &g * &a1;
+        // 2. t <- g * t'
+        let t = &g * &t1;
+        // 3. if t = 0 then return (a,a)
+        if t.is_zero() {
+            return Some(Self {
+                a: a.clone(),
+                b: a,
+                delta,
+            });
+        }
+        // 4. x <- t^2 * ∆ (mod a)
+        let t2 = BigInt::mod_mul(&t, &t, &a);
+        let x = BigInt::mod_mul(&t2, &delta, &a);
+        // 5. s <- sqrt(x)
+        let s = x.sqrt();
+        // 6. s' <- s/g
+        let s1 = &s / &g;
+        // 7. b' <- s' * t^−1 (mod a')
+        let t_inv = t.invert(&a1)?;
+        let b1 = BigInt::mod_mul(&s1, &t_inv, &a1);
+        // 8. b <- CRT((b', a'), (b0, g))
+        let mut b: BigInt =
+            ring_algorithm::chinese_remainder_theorem(&[A(b1), A(b0)], &[A(a1), A(g)])?
+                .into_inner();
+        // 9. if ε = False then b <- −b (mod a)
+        if !e {
+            b = -b
+        }
+        // 10. return (a,b)
+        Some(Self { a, b, delta })
+    }
+}
+
+/// Takes a,b (a > b > 0), produces r,s,t such as `r = s * a + t * b` where `|r|,|t| < sqrt(a)`
+fn partial_xgcd(a: &BigInt, b: &BigInt) -> (BigInt, BigInt, BigInt) {
+    let mut r = (a.clone(), b.clone());
+    let mut s = (BigInt::one(), BigInt::zero());
+    let mut t = (BigInt::zero(), BigInt::one());
+
+    let a_sqrt = a.sqrt();
+    while r.1 > a_sqrt {
+        let q = &r.0 / &r.1;
+        let r1 = &r.0 - &q * &r.1;
+        let s1 = &s.0 - &q * &s.1;
+        let t1 = &t.0 - &q * &t.1;
+
+        swap(&mut r.0, &mut r.1);
+        r.1 = r1;
+        swap(&mut s.0, &mut s.1);
+        s.1 = s1;
+        swap(&mut t.0, &mut t.1);
+        t.1 = t1;
+    }
+
+    (r.1, s.1, t.1)
+}
+
+/// Takes a, b, produces gcd(a,b), s, t such as `gcd(a,b) = s * a + t * b`
+fn xgcd(a: &BigInt, b: &BigInt) -> (BigInt, BigInt, BigInt) {
```

Comment:
can this be used instead: https://github.com/ZenGo-X/curv/blob/master/src/arithmetic/big_gmp.rs#L139 ?

Reply:
Sure! Thanks for link, didn't see it before.

Btw, I've discovered [a crate](https://docs.rs/ring-algorithm/0.2.3/ring_algorithm/) that generically implements some algorithms (gcd, extended gcd, chinese remainder theorem) which can be reused for curv::BigInt

Reply:
Just pushed a commit with fix